pkg/controller/infinispan/infinispan_controller.go

```diff
@@ -336,6 +334,18 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
 		return reconcile.Result{}, err
 	}
 
+	// Recover Pods with updated init containers in case of fails
+	for _, pod := range podList.Items {
+		if !kube.IsInitContainersEqual(statefulSet.Spec.Template.Spec.InitContainers, pod.Spec.InitContainers) {
+			if kube.IsInitContainerTerminated(pod.Status.InitContainerStatuses) {
+				if err = r.client.Delete(context.TODO(), &pod); err != nil {
+					return reconcile.Result{}, err
+				}
+				continue
+			}
+		}
+	}
```

Comment:
Having the init container retry logic in the operator with no upper bound will result in the pods being created/deleted repeatedly if it's not possible for a download to ever succeed (e.g. typo in the sha).

If we have the retry logic in the init container itself we can easily set an upper bound on the number of retries and we don't need any additional logic in the controller. If the download fails `x` times, then the initcontainer fails and the cluster can't be created.

Reply:
> Having the init container retry logic in the operator with no upper bound will result in the pods being created/deleted repeatedly if it's not possible for a download to ever succeed (e.g. typo in the sha).

Pods will never recreated if Init container is fail. Not sure we could embed retry logic into the download artifact script. I think that retry logic managed by k8s controller much better that own one.

Reply:
> Not sure we could embed retry logic into the download artifact script.

It's simple really, we just utilise a bash for loop e.g. `for i in 1 2 3 4 5; do command && break || sleep 1; done`

If `command` succeeds, then when don't retry as the exit code will be 0.

Reply:
Thanks. Will try to embed. Script logic become complex and complex :)

Reply:
Yes, good idea, but in this case we need to implement additional containers comparison logic. Lets try to use old approach.

Reply:
Also, I would prefer to make retry count and interval configurable, but we need to have Configuration CR first. Many project who start to keep configuration in the ConfigMap are moving away from this method to CRD as this is much flexible way.

Reply:
It would be good to have the retry count+interval configurable, but I don't think we need to worry too much about that unless there is demand from users.

+1 to a InfinispanConfiguration CR. That was my preference when the ConfigMap was initially introduced. I think this should probably wait until we have settled on a plan for CRD versioning.